cli-files/go-ls/cmd/root.go

```diff
@@ -0,0 +1,119 @@
+// cmd/root.go
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+)
+
+const (
+	ErrInvalidPath           = "Invalid path provided"
+	ErrNotADirectory         = "Provided path is not a directory"
+	ErrCouldNotReadDirectory = "Could not read directory"
+)
+
+var (
+	Delimeter string = "\n"
+	flagM     bool
+)
+
+func Execute() {
+	if err := golsCmd.Execute(); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+}
+
+var golsCmd = &cobra.Command{
+	Use:   "go-ls",
+	Short: "go-ls is an alternative to ls, allows you to list files in a given directory",
+	Long:  "go-ls is an alternative to ls, allows you to list files in a given directory. By default it lists files in the current working directory. You can also specify a directory to list files in.",
+	RunE:  GolsCmdFunc,
+}
+
+func init() {
+	golsCmd.Flags().BoolVarP(&flagM, "m", "m", false, "Use this flag if you would like the output to be comma separated instead of new lines.")
+}
+
+func GolsCmdFunc(cmd *cobra.Command, args []string) error {
+
+	var (
+		path string
+		err  error
+	)
+
+	if flagM {
+		Delimeter = ", "
+	}
+
+	switch len(args) {
+	case 0:
+		path, err = os.Getwd()
+		if err != nil {
+			return err
+		}
+		out, err := ListFolderContents(path)
+		if err != nil {
+			return err
+		}
+		fmt.Fprintln(cmd.OutOrStdout(), out)
+	case 1:
```

Comment:
Why does the 1 case need special handling  compared to the default case? I'd expect them to act the same? What's special about 1?

Reply:
It was based off of the behaviour for standard `ls` command. If it's more than one arg, then it adds the name of the path before printing out contents/error

Reply:
Cool, that makes sense!

One thing we often try to do when handling multiple similar cases is to structure our code to highlight what's the same and what's different. Right now, it's hard to tell what the difference is between the three cases here - to tell the difference, you need to read all three implementations, and carefully be spotting the differences.

Instead, we often try to structure our code as a single series of statements, with conditional logic to either do extra steps, or change variables.

For instance, let's look at these two ways of writing something:

```go
if x == 0 {
    r, s, err := call("abfjnkaeb")
    if err != nil {
        return fmt.Errorf("oh no: %w", err)
    }
    alsoCall(r, s)
} else if x == 1 {
    s, r, err := call("ABFJNKAEB1")
    if err != nil {
        return fmt.Errorf("oh no: %w", err)
    }
    alsoCall(r, s)
} else {
    r, s, err := call("abfjnkaeb2")
    if err != nil {
        return fmt.Errorf("oh no: %w", err)
    }
    alsoCall(r, s)
}
```

vs

```go
str := "abfjnkaeb"
if x == 1 {
    str = strings.ToUpper(str) + "1"
} else if x > 1 {
    str = str + "2"
}

r, s, err := call(str)
if err != nil {
    return fmt.Errorf("oh no: %w", err)
}

var first, second string
if x == 1 {
    first = s
    second = r
} else {
    first = r
    second = s
}
alsoCall(first, second)
```

These two blocks of code do the same thing, but in the first case there are a lot of subtle differences in each three blocks. Did you immediately spot all of them on your first read? Did you know the full differences in the strings passed to `call`? Did you notice that the arguments to `alsoCall` were swapped? How sure were you that the error handling was the same across all three branches? It takes a while to read three blocks of code and know what differences there are. It takes even longer to convince yourself that each of the differences are intentional! Are the arguments to `alsoCall` _meant_ to be swapped in the `x == 1` case? Was that a typo?

In the second one, we structure our code around what's the _same_ and what's _different_ - we can see that the error handling in each case is the same, because it's only there once. We can see what the relationship between the strings passed to `call` are because they're expressed in conditional logic. We can see that `r `and `s` _are_ indeed meant to be swapped - there's some very explicit logic saying when that should be. And if we come to extend the code in the future (let's say we add a special case for `x < 0`), we can write down the _differences_ that entails (maybe we don't call `alsoCall` in that case, so we'd add a conditional around it), rather than having to copy-paste again.

Can you try re-writing this `switch` block so that we have _one_ logical flow, with _small differences_ expressed, rather than having three copies of very similar logic?

Reply:
I have rewritten the whole block differently now. Hopefully it's less repetive and makes a bit more sense. Thank you!